h3/src/frame.rs

```diff
@@ -2,33 +2,38 @@ use std::task::{Context, Poll};
 
 use bytes::{Buf, Bytes};
 
+use futures::ready;
 use tracing::trace;
 
 use crate::{
     buf::BufList,
     error::TransportError,
-    proto::frame::{self, Frame},
+    proto::frame::{self, Frame, PayloadLen},
     quic::{RecvStream, SendStream},
+    stream::WriteBuf,
 };
 
 pub struct FrameStream<S>
 where
     S: RecvStream,
 {
     stream: S,
-    bufs: BufList<S::Buf>,
+    bufs: BufList<Bytes>,
```

Comment:
Just curious, why did this change from `S::Buf` to `Bytes`?

Reply:
In order to extract the next data chunk without consuming what's following it (either another `Data` header or the trailers) in the cheapest way possible, I chose to use `Bytes::split_to()`. As seen in :

```rust
impl BufList<Bytes> {
    pub fn take_chunk(&mut self, max_len: usize) -> Option<Bytes> {
        let chunk = self
            .bufs
            .front_mut()
            .map(|chunk| chunk.split_to(usize::min(max_len, chunk.remaining())));
            // [...]
    }
    // [...]
}
```

That would be harder or less efficient using only `Buf`. (I guess ?)

As we will call `<S::Buf as Buf>::copy_to_bytes()` at some point to return an owned "smart reference counted zero-copy slice" to the user, the best is to convert `RecvStream::Buf` into `Bytes` right after it's yeilded from the `quic` impl. So there are more chances it is a continuous chunk of bytes, and thus is cheaply convertible to `Bytes`.